lib/vrl/stdlib/src/parse_csv.rs

```diff
@@ -0,0 +1,127 @@
+use bytes::Buf;
+use csv::ReaderBuilder;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseCsv;
+
+impl Function for ParseCsv {
+    fn identifier(&self) -> &'static str {
+        "parse_csv"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "parse a single CSV formatted row",
+            source: r#"parse_csv!(s'foo,bar,"foo "", bar"')"#,
+            result: Ok(r#"["foo", "bar", "foo \", bar"]"#),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        let delimiter = arguments.optional("delimiter").unwrap_or(expr!(","));
+        Ok(Box::new(ParseCsvFn { value, delimiter }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "delimiter",
+                kind: kind::BYTES,
+                required: false,
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseCsvFn {
+    value: Box<dyn Expression>,
+    delimiter: Box<dyn Expression>,
+}
+
+impl Expression for ParseCsvFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let csv_string = self.value.resolve(ctx)?.try_bytes()?;
+        let delimiter = self.delimiter.resolve(ctx)?.try_bytes()?;
+        if delimiter.len() != 1 {
+            return Err("delimiter must be a single character".into());
+        }
+        let delimiter = delimiter[0];
+
+        let reader = ReaderBuilder::new()
+            .has_headers(false)
+            .delimiter(delimiter)
+            .from_reader(csv_string.bytes());
+
+        let result = reader
+            .into_byte_records()
+            .next()
+            .transpose()
+            .map_err(|err| format!("invalid csv record: {}", err))? // shouldn't really happen
+            .map(|record| record.iter().map(Into::into).collect::<Vec<Value>>())
+            .unwrap_or_default()
+            .into();
+
+        Ok(result)
```

Comment:
Nice! This looks cleaner.

Just a nit, but I think you could make this slightly more concise with:

```suggestion
        reader
            .into_byte_records()
            .next()
            .transpose()
            .map_err(|err| format!("invalid csv record: {}", err).into()) // shouldn't really happen
            .map(|record| {
                record
                    .iter()
                    .map(|r| r.as_slice().into())
                    .collect::<Vec<Value>>()
                    .into()
            })
```

To avoid the temporary variable.

Reply:
unfortunately `as_slice` here returns a whole byte array.

Reply:
Ah maybe you could just leave that as `.map(Into::into)` then.

Reply:
If I understood you correctly, it seems impossible because `r` here is of type `&ByteRecord`.

Reply:
Aha, yeah, I missed that `record` is an `Option`.

I think something like this then:

```rust
        reader
            .into_byte_records()
            .next()
            .transpose()
            .map_err(|err| format!("invalid csv record: {}", err).into()) // shouldn't really happen
            .map(|record| {
                record
                    .map(|record| record.iter().map(Into::into).collect::<Vec<Value>>())
                    .unwrap_or_default()
                    .into()
            })
```

Reply:
nice 👍  I was struggling with something like this because I didn't understand that this `into()` here `|err| format!("invalid csv record: {}", err).into()` is important. I updated it as you suggested. Also now I guess(please let me know if I'm mistaken) we have to switch back to `map_err` and can't use `expect` since we need to convert to `Resolved`, right?

Reply:
👍 You could leave the `expect` if you want. You just have to wrap the result in `Ok` like:

```rust
        Ok(reader
            .into_byte_records()
            .next()
            .transpose()
            .expect("unexpected error parsing csv")
            .map(|record| record.iter().map(Into::into).collect::<Vec<Value>>())
            .unwrap_or_default()
            .into())
```

Mapping errors with `into` trips me up too sometimes. The short-circuit `?` operator calls `into()` automatically.

Reply:
To be honest, I like bubbling up the error here instead anyway.

Reply:
Thanks a lot, it was a good learning exercise :)  Yeah, I'd leave it with bubbling up the error as well.  
Just to be sure, do you have any other comments or I can merge it?

Reply:
No problem! I would add this error to the list of `internal_failure_reasons` in the docs, but otherwise I think this is good to go.

Reply:
Right, added 👍